Clock frequency support for radeon and APUs v2

trek00 opened the pull request:
This patch series adds the clock frequency reporting for radeon and display it on APUs too. Code was refactored to split device dependent parts to amdgpu.c and radeon.c. DRM version is now checked to better report errors to the user.

Tested on Linux with amdgpu and radeon, with 32 and 64 bit builds, on KAVERI APU.

There are two unresolved caveats, because the maximum memory clock is not available when using radeon and the current memory clock is not available on APUs when using amdgpu. It will show an empty graph labeled like "0.80G / 0.00G Memory Clock inf%" or "0.00G / 0.80G Memory Clock 0.00%". May be this line should be disabled in these cases or is this useful anyway?

ciao!

clbr commented:
- why the amdgpu libdrm comment change? You yourself increased the needed version
- needless wrapping change in README
- please do not edit the .1 man page, that is generated from the .asc file
- for all those ifdef checks, did you test it compiles if such are not defined?
- it is not useful not to die if amdgpu init fails. No user is interested in a display of zeroes
- did you test everything fits on a default-sized terminal (80x24)?

trek00 commented:
> - why the amdgpu libdrm comment change? You yourself increased the
> needed version

after looking the libdrm git, I discovered that amdgpu.h was introduced
in version 2.4.63 with grbm and vram support, while its clock sensor
support was added in 2.4.79, thus the minimal version to compile with
amdgpu is 2.4.63, but it will get clock readings only if compiled with
libdrm >= 2.4.79

commit 2c32d added a pkg-config check to libdrm_amdgpu for grbm and
vram support, then a second check for 2.4.79 version, to add a define
named HAS_AMDGPU_QUERY_SENSOR_INFO that is used to guard clock sensor
code in detect.c (now in amdgpu.c)

> - needless wrapping change in README

I'll revert it after fixing other issues and rebasing

> - please do not edit the .1 man page, that is generated from the .asc
> file

thanks I thought the opposite

> - for all those ifdef checks, did you test it compiles if such are
> not defined?

yes, I tested every ifdef and run-time check at least once

> - it is not useful not to die if amdgpu init fails. No user is
> interested in a display of zeroes

just in case amdgpu kernel driver or libdrm are not working fine, may
be the /dev/mem path could be used as fallback to display at least grbm

> - did you test everything fits on a default-sized terminal (80x24)?

yes, it fits and I haven't added new code to ui.c, the only change was
a fix to display clocks on APUs too

many thanks for your comments and I hope this answer your questions
ciao!

trek00 commented:
I pushed a fixed and rebased version on new branch [trekk00:drm2](https://github.com/trek00/radeontop/tree/drm2).

To ease the review, here a list of changes:
- reverted indentation change from README.md
- modified radeontop.asc instead of radeontop.1
- removed RADEON_INFO_* defines as they can defeat autodetect of libdrm features
- drm_fd and drm_name are no more globals in detect.c
- static global variables in radeon.c and amdgpu.c
- getgrbm_device_mem renamed to getgrbm_mem instead of getgrbm_memory

The output could not fit 24 lines on old cards, as it displays more fields than with newer cards. I think ui.c could be reworked, but it is beyond the scope of these patches. Actually I'm working to add gpu detection via drmGetDevices, as it should fix bugs 36, 48 and 72.

Please to tell me if other changes or further discussion are needed.
ciao!

clbr commented:
Just a warning, I'm going to be out for a few weeks.

clbr commented:
drm2 is merged, thanks.
